### PR TITLE
Add RESTClient: 

### PR DESCRIPTION
**Pull request by burnash**
### Description

This pull request introduces a `RESTClient` class along with supporting components to interact with RESTful APIs. The RESTClient provides:
- an interface for making HTTP requests, 
- handling pagination (via a set of paginator classes: e.g. `HeaderLinkPaginator`, `JSONResponsePaginator`.
- managing authentication (via `BearerTokenAuth`, `HttpBasicAuth`, etc.)

Users are able to customize request handling, response parsing, and data extraction according to the specifics of the target API. `RESTClient` also features automatic paginator detection with the `PaginatorFactory`

Example usage:

```python
import dlt
from dlt.sources.helpers.rest_client import RESTClient

github_client = RESTClient(base_url="https://api.github.com")

@dlt.resource(
    table_name="issues",
    write_disposition="merge",
    primary_key="id",
)
def get_issues(
    updated_at=dlt.sources.incremental(
        "updated_at", initial_value="1970-01-01T00:00:00Z"
    )
):
    for page in github_client.paginate(
        "/repos/dlt-hub/dlt/issues",
        params={
            "since": updated_at.last_value,
            "per_page": 100,
            "sort": "updated",
            "direction": "desc",
            "state": "open",
        },
    ):
        yield page


# The rest of the pipeline remains the same
pipeline = dlt.pipeline(
    pipeline_name="github_issues_merge",
    destination="duckdb",
    dataset_name="github_data_merge",
)
load_info = pipeline.run(get_issues)
print(load_info)
```

### Related Issues

- This is an extract from the `rest_api` source from https://github.com/dlt-hub/verified-sources/pull/313

### CODE REVIEW

**Comment by burnash** [comment]
@rudolfix I've put it in `dlt.sources.helpers.rest_client` module. Open for suggestions.

**Comment by rudolfix** [comment]
@burnash what I think is missing for internal testing is a simplified `request` like interface to access the rest client. the idea is to expose it from request helper.

```python
from source.helpers import requests

requests.paginate(url, method, auth=str | AuthBase | AuthBase(), paginator = str | Paginator | Paginator(), ....)
```

auth and paginators follow the same convention as destinations, progress, naming convention

str - a short hand alias ie or type
type or instance of a type
types have aliases identical to short hand ie `basic` is basic auth shorthand alias and a type. basic(user, password) creates BasicAuthenticator

auth and paginators can be configured

```toml
[source.chargebee.paginator]
cursor_field="..."
cursor_param="cursor"

[source.chargebee.auth]
api_key="...."
```

where to put REST client: current location is good. let's write some code and maybe we'll change it
I'll review the actual code tomorrow

File: dlt/sources/helpers/rest_client/utils.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+    # Normalize the base URL
+    base_url = base_url.rstrip("/")
+    if not base_url.endswith("/"):
+        base_url += "/"
+
+    return base_url + path.lstrip("/")
+
+
+def check_connection(
```
Comment (rudolfix) [inline dlt/sources/helpers/rest_client/utils.py:25]: I'd remove those and keep in verified sources. WDYT?

**Review by rudolfix — COMMENTED** [review]
this is good! we are missing
1. tests for requests.paginated shorthand
2. some initial documentation (maybe page in docs but not linked to index yet)

**Review by rudolfix — DISMISSED** [review]
LGTM! let' work on it in followup PRs

File: dlt/sources/helpers/rest_client/utils.py (hunk 1 above)
Comment (burnash) [reply dlt/sources/helpers/rest_client/utils.py:25]: Yes, it's here by mistake

**Comment by burnash** [comment]
@rudolfix 

> 1. tests for requests.paginated shorthand

There are tests for that in the `test_client.py` but I think it's best to have a separate file for that.

File: dlt/sources/helpers/rest_client/utils.py (hunk 1 above)
Comment (burnash) [reply dlt/sources/helpers/rest_client/utils.py]: Fixed in 9b3d353b9728c1bbf3608135ecf6f5cc6e70cf3a

**Comment by burnash** [comment]
@rudolfix

> tests for requests.paginated shorthand

Refactored from test_client.py in b1ee49ab10c4426f1aa4a9d12b6d8bdb77a42a73

File: dlt/sources/helpers/rest_client/auth.py [hunk 2]
```diff
@@ -104,16 +88,8 @@ def __call__(self, request: PreparedRequest) -> PreparedRequest:
...
-    scheme: Literal["basic"] = "basic"
-    username: str
-    password: TSecretStrValue
-
-    def __init__(
-        self, username: str = config.value, password: TSecretStrValue = secrets.value
-    ) -> None:
-        self.username = username
-        self.password = password
+    username: str = ""
```
Comment (rudolfix) [inline dlt/sources/helpers/rest_client/auth.py:91]: it is better to set `username` to None. this makes it a required config field. "" makes it optional (None is used to detect that field is missing)

**Review by rudolfix — CHANGES_REQUESTED** [review]
pls see one comment

File: dlt/sources/helpers/rest_client/auth.py (hunk 2 above)
Comment (burnash) [reply dlt/sources/helpers/rest_client/auth.py]: Fixed in b9bb0d354aa281fb84838f39d7e3f25e7fe9b040

**Review by rudolfix — COMMENTED** [review]
LGTM!

**Pull request merged**